cpp17/diff_match_patch.cpp

```diff
@@ -0,0 +1,2200 @@
+        break;
+      case EQUAL:
+        text += L"=" + std::to_wstring(aDiff.text.length()) + L"\t";
+        break;
+    }
+  }
+  if (!text.empty()) {
+    // Strip off trailing tab character.
+    text = text.substr(0, text.length() - 1);
+  }
+  return text;
+}
+
+TDiffVector diff_match_patch::diff_fromDelta(const std::wstring &text1,
+                                             const std::wstring &delta) {
+  TDiffVector diffs;
+  std::size_t pointer = 0;  // Cursor in text1
+  auto tokens = NUtils::splitString(delta, L"\t", false);
+  for (auto &&token : tokens) {
+    if (token.empty()) {
+      // Blank tokens are ok (from a trailing \t).
+      continue;
+    }
+    // Each token begins with a one character parameter which specifies the
+    // operation of this token (delete, insert, equality).
+    std::wstring param = safeMid(token, 1);
+    switch (token[0]) {
+      case '+':
+        NUtils::replace(param, L"+", L"%2b");
+        param = NUtils::fromPercentEncoding(param);
+        diffs.emplace_back(INSERT, param);
+        break;
+      case '-':
+        // Fall through.
+      case '=': {
+        auto n = NUtils::toInt(param);
+        if (n < 0) {
+          throw std::wstring(L"Negative number in diff_fromDelta: " + param);
+        }
+        std::wstring text;
+        if ((pointer + n) > text1.length()) {
+          throw std::wstring(L"Delta length (" + std::to_wstring(pointer + n) +
+                             L") larger than source text length (" +
+                             std::to_wstring(text1.length()) + L").");
+        }
+
+        text = safeMid(text1, pointer, n);
+        pointer += n;
+        if (token[0] == L'=') {
+          diffs.emplace_back(EQUAL, text);
+        } else {
+          diffs.emplace_back(DELETE, text);
+        }
+        break;
+      }
+      default:
+        throw std::wstring(L"Invalid diff operation in diff_fromDelta: " +
+                           token[0]);
+    }
+  }
+  if (pointer != text1.length()) {
+    throw std::wstring(L"Delta length (") + std::to_wstring(pointer) +
+        L") smaller than source text length (" +
+        std::to_wstring(text1.length()) + L")";
+  }
+  return diffs;
+}
+
+TDiffVector diff_match_patch::diff_fromDelta(const std::string &text1,
+                                             const std::string &delta) {
+  return diff_fromDelta(NUtils::to_wstring(text1), NUtils::to_wstring(delta));
+}
+
+//  MATCH FUNCTIONS
+
+std::size_t diff_match_patch::match_main(const std::wstring &text,
+                                         const std::wstring &pattern,
+                                         std::size_t loc) {
+  // Check for null inputs not needed since null can't be passed via
+  // std::wstring
+
+  loc = std::max(kZERO, std::min(loc, text.length()));
+  if (text == pattern) {
+    // Shortcut (potentially not guaranteed by the algorithm)
+    return 0;
+  } else if (text.empty()) {
+    // Nothing to match.
+    return -1;
```

Comment:
Warning	C4245	'return': conversion from 'int' to 'size_t', signed/unsigned mismatch

Maybe use `return std::wstring::npos;` isntead?